Readme.md

# phony

  Tiny command line program that accepts a template and outputs fake data.

  ![](https://cldup.com/RZoAhReDqN.gif)

## Installation

Via gobinaries.com

```bash
$ curl https://gobinaries.com/yields/phony | sh
```

Via go get:

```bash
$ go get github.com/yields/phony
```

## Examples

```bash
# publish email to nsq every 1ms.
echo '{"email":"{{email}}", "subject": "welcome!"}' \
  | phony --tick 1ms \
  | json-to-nsq --topic users

# add users to FoundationDB.
echo "'set {{username}} {{avatar}}'" \
  | phony \
  | xargs -L1 -n3 fdbcli --exec

# add users to MongoDB.
echo "'db.users.insert({ name: \"{{name}}\" })'" \
  | phony \
  | xargs -L1 -n1 mongo --eval

# add users to Redis.
echo "set {{username}} {{avatar}}" \
  | phony \
  | xargs -L1 -n3 redis-cli

# send a single request using curl.
echo 'country={{country}}' \
  | phony --max 1 \
  | curl -d @- httpbin.org/post
```

## Usage

```text

Usage: phony
  [--tick d]
  [--max n]
  [--list]

  phony -h | --help
  phony -v | --version

Options:
  --list          list all available generators
  --max n         generate data up to n [default: -1]
  --tick d        generate data every d [default: 10ms]
  -v, --version   show version information
  -h, --help      show help information

```

## Generators

```text
  avatar
  color
  country
  country.code
  domain
  domain.name
  domain.tld
  double
  email
  event.action
  http.method
  id
  ipv4
  ipv6
  ksuid
  latitude
  longitude
  mac.address
  name
  name.first
  name.last
  product.category
  product.name
  state
  state.code
  timezone
  unixtime
  username
  uuid
```

## License

  (MIT), 2014 Amir Abu Shareb.


History.md

v0.0.1 / 2020-03-28
===================

  * Initial release


pkg/phony/phony_test.go
package phony

import "github.com/bmizerany/assert"
import "testing"

func TestGet(t *testing.T) {
	a, _ := Get("name")
	b, _ := Get("name")
	assert.NotEqual(t, a, "")
	assert.NotEqual(t, b, "")
}

func TestEmpty(t *testing.T) {
	a, _ := Get("foo")
	assert.Equal(t, a, "")
}

func TestAll(t *testing.T) {
	for _, p := range List() {
		a, _ := Get(p)
		assert.NotEqual(t, a, "")
	}
}


pkg/phony/helpers_test.go
package phony

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestStringsToFloatParsesStringsAsFloats(t *testing.T) {
	a := []string{"10", "-100.1", "1e-3", "-1e3"}

	b := stringsToFloats(a)

	assert.Equal(t, []float64{10.0, -100.1, 0.001, -1000}, b)
}

func TestStringsToFloatStopsParsingOnFirstFailure(t *testing.T) {
	a := []string{"10", "abc", "1e-3", "-1e3"}

	b := stringsToFloats(a)

	assert.Equal(t, []float64{10.0}, b)
}


main.go
package main

import "github.com/yields/phony/pkg/phony"
import "github.com/tj/docopt"
import "math/rand"
import "io/ioutil"
import "strconv"
import "strings"
import "regexp"
import "sort"
import "time"
import "fmt"
import "os"

var usage = `
  Usage: phony
    [--tick d]
    [--max n]
    [--list]

    phony -h | --help
    phony -v | --version

  Examples:

    # output names
    echo '{{ name }}' | phony

    # output names every 1s
    echo '{{ name }}' | phony --tick 1s

    # output a sigle name
    echo '{{ name }}' | phony --max 1

  Options:
    --list          list all available generators
    --max n         generate data up to n [default: -1]
    --tick d        generate data every d [default: 10ms]
    -v, --version   show version information
    -h, --help      show help information

`

func main() {
	args, err := docopt.Parse(usage, nil, true, "0.0.1", false)
	check(err)

	if args["--list"].(bool) {
		all := phony.List()
		sort.Strings(all)
		println()
		for _, name := range all {
			fmt.Printf("  %s\n", name)
		}
		println()
		os.Exit(0)
	}

	rand.Seed(time.Now().UnixNano())

	d := parseDuration(args["--tick"].(string))
	max := parseInt(args["--max"].(string))

	if 0 >= d {
		fmt.Fprintf(os.Stderr, "phony: --tick must be a positive interval, got %q\n", d)
		os.Exit(1)
	}

	tmpl := readAll(os.Stdin)

	ticker := time.NewTicker(d)
	defer ticker.Stop()
	f := compile(string(tmpl))
	it := 0

	for range ticker.C {
		fmt.Fprintf(os.Stdout, "%s", f())
		if it++; -1 != max && it == max {
			return
		}
	}
}

func compile(tmpl string) func() string {
	expr, err := regexp.Compile(`({{ *(([a-zA-Z0-9]+(\.[a-zA-Z0-9]+)?)+(\:([a-zA-Z0-9,]+))?) *}})`)
	check(err)
	return func() string {
		return expr.ReplaceAllStringFunc(tmpl, func(s string) string {
			call := strings.Trim(s[2:len(s)-2], " ")
			parts := strings.Split(call, ":")
			var arguments []string = nil
			if len(parts) == 2 {
				arguments = strings.Split(parts[1], ",")
			}
			data, err := phony.GetWithArgs(parts[0], arguments)
			check(err)
			return data
		})
	}
}

func check(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "phony: %s\n", err.Error())
		os.Exit(1)
	}
}

func parseInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}

func parseDuration(s string) time.Duration {
	d, err := time.ParseDuration(s)
	check(err)
	return d
}

func readAll(r *os.File) string {
	b, err := ioutil.ReadAll(r)
	check(err)
	return string(b)
}


pkg/phony/phony.go
package phony

import "math/rand"

// Default generator.
var gen = New(&Dataset{
	gens: gens,
	dict: dict,
})

// Dataset.
type Dataset struct {
	gens map[string]func(g *Generator, args []string) (string, error)
	dict map[string][]string
}

// Generator structure.
type Generator struct {
	set *Dataset
}

// Initialize Generator with `dataset`.
func New(set *Dataset) *Generator {
	return &Generator{set}
}

// Get `path`.
func (g *Generator) Get(p string) (string, error) {
	return g.GetWithArgs(p, nil)
}

// Get `path`.
func (g *Generator) GetWithArgs(p string, args []string) (string, error) {
	gens := g.set.gens
	dict := g.set.dict

	if f, ok := gens[p]; ok {
		return f(g, args)
	}

	if list, ok := dict[p]; ok {
		i := rand.Intn(len(list))
		return list[i], nil
	}

	return "", nil
}

// List all paths.
func (g *Generator) List() []string {
	gens := g.set.gens
	dict := g.set.dict
	ret := make([]string, 0)

	for k, _ := range gens {
		ret = append(ret, k)
	}

	for k, _ := range dict {
		ret = append(ret, k)
	}

	return ret
}

// Get `path`.
func Get(path string) (string, error) {
	return gen.Get(path)
}

// Get `path`.
func GetWithArgs(path string, args []string) (string, error) {
	return gen.GetWithArgs(path, args)
}

// List all available paths.
func List() []string {
	return gen.List()
}


pkg/phony/generator.go
package phony

import (
	"errors"
	"fmt"
	"math/rand"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	smartdouble   = "smartdouble"
	smartunixtime = "smartunixtime"
	smartdate     = "smartdate"
)

var supportedDateFormats = map[string]string{
	// Go default formats
	"ANSIC":       "Mon Jan _2 15:04:05 2006",
	"UnixDate":    "Mon Jan _2 15:04:05 MST 2006",
	"RubyDate":    "Mon Jan 02 15:04:05 -0700 2006",
	"RFC822":      "02 Jan 06 15:04 MST",
	"RFC822Z":     "02 Jan 06 15:04 -0700",
	"RFC850":      "Monday, 02-Jan-06 15:04:05 MST",
	"RFC1123":     "Mon, 02 Jan 2006 15:04:05 MST",
	"RFC1123Z":    "Mon, 02 Jan 2006 15:04:05 -0700",
	"RFC3339":     "2006-01-02T15:04:05Z07:00",
	"RFC3339Nano": "2006-01-02T15:04:05.999999999Z07:00",
	"Kitchen":     "3:04PM",
	"Stamp":       "Jan _2 15:04:05",
	"StampMilli":  "Jan _2 15:04:05.000",
	"StampMicro":  "Jan _2 15:04:05.000000",
	"StampNano":   "Jan _2 15:04:05.000000000",
	// Additional formats
	"SqlDatetime": "2006-01-02 15:04:05",
	"SqlDate":     "2006-01-02",
	"SqlTime":     "15:04:05",
}

type Generator struct {
	sync.Mutex
	cachedArguments map[string]map[string]interface{}
	argumentGetters map[string]func(g *Generator, args []string) interface{}
	generators      map[string]func(g *Generator, args []string, r *rand.Rand) (string, error)
	dictionary      map[string][]string
}

func (g *Generator) getSmartDate(args []string, r *rand.Rand) time.Time {
	date := time.Now()

	floatArgs := g.argumentGetters[smartunixtime](g, args).([]float64)

	// Deviation
	if len(floatArgs) > 0 {
		addDays := int64(floatArgs[0] * float64(time.Hour) * 24.0)
		date = date.Add(time.Duration(addDays))
	}

	// Scatter
	if len(floatArgs) > 1 {
		scatterFactor := (r.Float64() - 0.5) * 2
		scatterDays := int64(floatArgs[1] * float64(time.Hour) * 24.0 * scatterFactor)

		date = date.Add(time.Duration(scatterDays))
	}

	return date
}

var gen *Generator

func NewGenerator() *Generator {
	if gen != nil {
		return gen
	}

	var generator = Generator{}

	generator.argumentGetters = map[string]func(g *Generator, args []string) interface{}{
		smartdouble: func(g *Generator, args []string) interface{} {
			key := strings.Join(args, ",")

			if found, ok := g.cachedArguments[smartdouble][key]; ok {
				return found
			}

			g.Lock()
			defer g.Unlock()
			g.cachedArguments[smartdouble][key] = stringsToFloats(args)

			return g.cachedArguments[smartdouble][key]
		},
		smartunixtime: func(g *Generator, args []string) interface{} {
			key := strings.Join(args, ",")

			if found, ok := g.cachedArguments[smartdouble][key]; ok {
				return found
			}

			g.Lock()
			defer g.Unlock()
			g.cachedArguments[smartunixtime][key] = stringsToFloats(args)

			return g.cachedArguments[smartunixtime][key]
		},
	}

	generator.cachedArguments = map[string]map[string]interface{}{
		smartdouble:   map[string]interface{}{},
		smartunixtime: map[string]interface{}{},
	}

	// Default gens.
	generator.generators = map[string]func(g *Generator, args []string, r *rand.Rand) (string, error){
		"name": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			a, _ := g.Get("name.first", r)
			b, _ := g.Get("name.last", r)
			return a + " " + b, nil
		},
		"email": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			username, _ := g.Get("username", r)
			host, _ := g.Get("domain", r)
			return fmt.Sprintf("%s%d@%s", username, r.Intn(253), host), nil
		},
		"domain": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			name, _ := g.Get("domain.name", r)
			tld, _ := g.Get("domain.tld", r)
			return name + "." + tld, nil
		},
		"avatar": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			// http://uifaces.com/authorized
			user, _ := g.Get("username", r)
			return "https://s3.amazonaws.com/uifaces/faces/twitter/" + user + "/128.jpg", nil
		},
		"unixtime": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			return strconv.FormatInt(time.Now().UnixNano(), 10), nil
		},
		"id": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			chars := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
			ret := make([]rune, 10)

			for i := range ret {
				ret[i] = chars[r.Intn(len(chars))]
			}

			return string(ret), nil
		},
		"ipv4": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			return fmt.Sprintf("%d.%d.%d.%d", 1+r.Intn(253), r.Intn(255), r.Intn(255), 1+r.Intn(253)), nil
		},
		"ipv6": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			return fmt.Sprintf("2001:cafe:%x:%x:%x:%x:%x:%x", r.Intn(255), r.Intn(255), r.Intn(255), r.Intn(255), r.Intn(255), r.Intn(255)), nil
		},
		"mac.address": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			return fmt.Sprintf("%x:%x:%x:%x:%x:%x", r.Intn(255), r.Intn(255), r.Intn(255), r.Intn(255), r.Intn(255), r.Intn(255)), nil
		},
		"latitude": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			lattitude := (r.Float64() * 180) - 90
			return strconv.FormatFloat(lattitude, 'f', 6, 64), nil
		},
		"longitude": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			longitude := (r.Float64() * 360) - 180
			return strconv.FormatFloat(longitude, 'f', 6, 64), nil
		},
		"double": func(g *Generator, args []string, r *rand.Rand) (string, error) {
			return strconv.FormatFloat(r.NormFloat64()*1000, 'f', 4, 64), nil
		},
		// Smartdouble returns a random double
		// First argument will be interpreted as desired standard deviation
		// Second argument will be interpreted as desired mean
		// Third argument will be interpreted as minimum value expected
		// Fourth argument will be interpreted as maximum value expected
		//
		// More info:
		// https://golang.org/pkg/math/rand/#Rand.NormFloat64
		smartdouble: func(g *Generator, args []string, r *rand.Rand) (string, error) {
			var (
				desiredStdDev = 1000.0
				desiredMean   = 0.0
			)

			floatArgs := g.argumentGetters[smartdouble](g, args).([]float64)

			if len(floatArgs) > 0 {
				desiredStdDev = floatArgs[0]
			}

			if len(floatArgs) > 1 {
				desiredMean = floatArgs[1]
			}

			randNum := r.NormFloat64()*desiredStdDev + desiredMean

			if len(floatArgs) > 2 && randNum < floatArgs[2] {
				randNum = floatArgs[2]
			}

			if len(floatArgs) > 3 && randNum > floatArgs[3] {
				randNum = floatArgs[3]
			}

			return strconv.FormatFloat(randNum, 'f', 4, 64), nil
		},
		// Smartunixtime returns a random unix time based on the current time, args can be an array of strings.
		// Each string is expected to be float64 parsable. First argument will be interpreted as deviation days,
		// second argument will be interpreted as days for allowed range in days.
		//
		// Example:
		//   Today is: 2006-01-02T15:04:05Z07:00
		//   Args: ["10", "5"]
		//     Deviation is: 10 (days)
		//     Scatter is: 5 (days)
		//   Result: unix timestamp representing date between 2006-01-07T15:04:05Z07:00 and 2006-01-17T15:04:05Z07:00
		smartunixtime: func(g *Generator, args []string, r *rand.Rand) (string, error) {
			date := g.getSmartDate(args, r)

			return strconv.FormatInt(date.UnixNano(), 10), nil
		},
		// Smartdate returns a random date based on the current time, args can be an array of strings.
		// First argument will be interpreted as expected date format
		// Second argument will be interpreted as deviation days,
		// Third argument will be interpreted as days for allowed range in days.
		//
		// Example:
		//   Today is: 2006-01-02T15:04:05Z07:00
		//   Args: ["SQL_DATE", "10", "5"]
		//     Deviation is: 10 (days)
		//     Scatter is: 5 (days)
		//   Result: string date between 2006-01-07 and 2006-01-17
		smartdate: func(g *Generator, args []string, r *rand.Rand) (string, error) {
			var (
				format       = "RFC3339"
				unixtimeArgs = []string{}
				ok           bool
			)

			if len(args) > 0 {
				format = args[0]
				unixtimeArgs = args[1:]
			}

			if format, ok = supportedDateFormats[format]; !ok {
				return "", errors.New("Invalid date format.")
			}

			date := g.getSmartDate(unixtimeArgs, r)

			return date.Format(format), nil
		},
	}

	generator.dictionary = dict

	gen = &generator

	return gen
}


pkg/phony/gens.go
package phony

import (
	"fmt"
	"math/rand"
	"strconv"
	"time"

	"github.com/google/uuid"
	"github.com/segmentio/ksuid"
)

// Default gens.
var gens = map[string]func(g *Generator, args []string) (string, error){
	"now.utc": func(g *Generator, args []string) (string, error) {
		return time.Now().UTC().Format(time.RFC3339), nil
	},
	"name": func(g *Generator, args []string) (string, error) {
		a, _ := g.Get("name.first")
		b, _ := g.Get("name.last")
		return a + " " + b, nil
	},
	"email": func(g *Generator, args []string) (string, error) {
		username, _ := g.Get("username")
		host, _ := g.Get("domain")
		return username + "@" + host, nil
	},
	"domain": func(g *Generator, args []string) (string, error) {
		name, _ := g.Get("domain.name")
		tld, _ := g.Get("domain.tld")
		return name + "." + tld, nil
	},
	"avatar": func(g *Generator, args []string) (string, error) {
		// http://uifaces.com/authorized
		user, _ := g.Get("username")
		return "https://s3.amazonaws.com/uifaces/faces/twitter/" + user + "/128.jpg", nil
	},
	"unixtime": func(g *Generator, args []string) (string, error) {
		return strconv.FormatInt(time.Now().UnixNano(), 10), nil
	},
	"id": func(g *Generator, args []string) (string, error) {
		chars := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
		ret := make([]rune, 10)

		for i := range ret {
			ret[i] = chars[rand.Intn(len(chars))]
		}

		return string(ret), nil
	},
	"uuid": func(g *Generator, args []string) (string, error) {
		return uuid.New().String(), nil
	},
	"ksuid": func(g *Generator, args []string) (string, error) {
		return ksuid.New().String(), nil
	},
	"ipv4": func(g *Generator, args []string) (string, error) {
		return fmt.Sprintf("%d.%d.%d.%d", 1+rand.Intn(253), rand.Intn(255), rand.Intn(255), 1+rand.Intn(253)), nil
	},
	"ipv6": func(g *Generator, args []string) (string, error) {
		return fmt.Sprintf("2001:cafe:%x:%x:%x:%x:%x:%x", rand.Intn(255), rand.Intn(255), rand.Intn(255), rand.Intn(255), rand.Intn(255), rand.Intn(255)), nil
	},
	"mac.address": func(g *Generator, args []string) (string, error) {
		return fmt.Sprintf("%x:%x:%x:%x:%x:%x", rand.Intn(255), rand.Intn(255), rand.Intn(255), rand.Intn(255), rand.Intn(255), rand.Intn(255)), nil
	},
	"latitude": func(g *Generator, args []string) (string, error) {
		lattitude := (rand.Float64() * 180) - 90
		return strconv.FormatFloat(lattitude, 'f', 6, 64), nil
	},
	"longitude": func(g *Generator, args []string) (string, error) {
		longitude := (rand.Float64() * 360) - 180
		return strconv.FormatFloat(longitude, 'f', 6, 64), nil
	},
	"double": func(g *Generator, args []string) (string, error) {
		return strconv.FormatFloat(rand.NormFloat64()*1000, 'f', 4, 64), nil
	},
}
